python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,263 @@
+from time import sleep
+from timeit import default_timer as clock
+from collections import deque
+from MooseObject import MooseObject
+from Queue import Queue
+from signal import SIGTERM
+from util import *
+import os, sys
+
+from multiprocessing.pool import ThreadPool
+import threading # for thread locking
+
+class Scheduler(MooseObject):
+    """
+    Base class for handling how jobs are launched. To use this class, call .schedule()
+    and supply a list of testers to schedule.
+
+    Syntax:
+       .schedule([list of tester objects])
+
+    A list of testers will be added to a queue and begin calling the derived run method
+    immediately. You can continue to add more testers to the queue in this fashion.
+
+    Once you schedule all the testers you wish to test, call .waitFinish() to wait until
+    all tests have finished.
+
+    As tests finish individually, the scheduler will call back handleTestStatus(tester)
+    """
+
+    @staticmethod
+    def validParams():
+        params = MooseObject.validParams()
+        params.addRequiredParam('average_load',          64.0, "Average load to allow")
+        params.addRequiredParam('max_processes',         None, "Hard limit of maxium processes to use")
+
+        return params
+
+    ## Return this return code if the process must be killed because of timeout
+    TIMEOUT = -999999
+
+    def __init__(self, harness, params):
+        MooseObject.__init__(self, harness, params)
+
+        self.specs = params
+
+        ## The test harness to run callbacks on
+        self.harness = harness
+
+        # Retrieve and store the TestHarness options for use in this object
+        self.options = harness.getOptions()
+
+        # Set max jobs to allow
+        params['max_processes'] = self.options.jobs
+        params['average_load'] = self.options.load
+
+        # For backwards compatibitliy the RunParallel class can be initialized
```

Comment:
backwards compatibility? I don't think we want to drop this design moving forward do we?